content/blog/2019-10-18-double-checked-locking/index.md

```diff
@@ -0,0 +1,353 @@
+    which have a huge performance impact.
+
+### Double-checked locking is broken 
+
+What we want is like following:
+
+1. check if the object is initialized without lock.
+If it is, then return the object immediately.
+2. acquire the lock and check again if the object is initialized.
+If another thread have grabbed lock before, 
+   the current thread can see the object is created, and return the object.
+3. otherwise, the current thread will create the object and return.
+
+With the guideline above, we will get a code like following:
+
+```Java
+class Foo {
+    private Helper helper = null;
+    public Helper getHelper() {
+        if (helper == null) {              // first check
+            synchronized(this) {
+                if (helper == null)         // second check
+                    helper = new Helper();
+            }
+        }
+        return helper;
+    }
+}
+```
+
+Intuitively, this is a simple fix, but this code is not guaranteed to work.
+Some compilers, memory systems, or processors may reorder the instructions,
+     so that the initialization of the object is done after the assignment of the pointer to the object.
+     Another thread may see that the pointer is assigned and access the fields of the object before the fields of the object is initialized.
+
+### Another fix is also broken
+
+Some people came up with another fix for this issue with a memory barrier.
+A memory barrier is a type of instruction that can make the compiler and processor
+enforce the ordering, so that the instructions on one side of the memory barrier will
+not be reordered to the other side of the barrier.
+
+```Java
+class Foo {
+    private Helper helper = null;
+
+    public Helper getHelper() {
+        if (helper == null) {
+            Helper h;
+            synchronized(this) {
+                h = helper;
+                if (h == null)
+                    synchronized(this) {
+                        h = new Helper();
+                    }                       
+                helper = h;
+            }
+        }
+        return helper;
+    }
+}
+
+```
+
+The intuitive idea is that the lock releasing would act as a memory barrier,
+so that `helper=h` will not be executed until the initialization 
+in the synchronized section is done.
+
+Unfortunately, the lock releasing is a one-way memory barrier on many processors.
+It only enforces that the instructions in the synchronized section 
+must be executed before lock releasing.
+The instruction `helper=h` behind the memory barrier could still be moved into 
+synchronized section and executed before the object initialization is done.
+
+## Working Solutions
+
+
+### Volatile in JDK5
+
+Java introduced "volatile" since JDK5. 
+A "volatile" makes the variable be stored and accessed in main memory.
+Every read of a volatile is from main memory.
+Every write of a volatile is to main memory.
+
+The "volatile" also provides a "happen-before" guarantee:
```

Comment:
Hmm, I don't think I'd use the term "happens-before"… it's too intimately tied to the Lamport ordering notion. https://lamport.azurewebsites.net/pubs/time-clocks.pdf